doc/dtc_write_xenpv_conf.8
.TH dtc_write_xenpv_conf 8
.SH NAME
dtc_write_xenpv_conf \- rewrites the xen startup script with selected params
.SH SYNOPSIS
.B dtc_write_xenpv_conf
.I VPS_NUMBER
.I MEMORY_SIZE
.I ALL_IP_ADDRESSES
.I VNC_CONSOLE_PASSWORD
.I HOW_TO_BOOT

.SH DESCRIPTION
.B dtc_write_xenpv_conf
This shell script is a part of the dtc\-xen
package that is to be used by the dtc panel
to manage a Xen VPS server.

This script generate a Xen startup file with the parameters
given to the script.

.SH PARAMETERS

Note that ALL parameters are mendatory.

.B VPS_NUMBER
has to be a number between 01 and 29

.B MEMORY_SIZE
is the amount of memory in mega bytes that you want to have setup
in the Xen startup file for this VPS

.B ALL_IP_ADDRESSES
is a list of IPs that you want to be set in the VM's startup file and
in the VM's configuration. Note that the IPs have to be separated by spaces,
and thus you might want to use single quotes like this: '1.2.3.4 1.2.3.5'. Note
that only the first IP will be set in the VM's configuration, but the list will
be set in the Xen startup file so you can use the anti\-spoofing facilities of
Xen if you need it (with untrusted users / unknown customers for example).

.B VNC_CONSOLE_PASSWORD
This is the VNC console password that you want to have set\-up in the Xen domU
configuration file, so that you can connect using VNC to the the setup. Note
that special value
.I no_vnc
can be used to declare that you do not want the VNC console to be used at all.

.B HOW_TO_BOOT
This controls what the VPS will use to boot. A special value of
.I hdd
can be used to tell dtc_write_xenpv_conf that you want to boot on the hard
drive of the domU. Any other parameters will be used as a .iso file located
in /var/lib/dtc\-xen/ttyssh_home/xenXX where xenXX equals to the VPS number
set in the
.I VPS_NUMBER
parameter (see above).

Note that whatever is set in this parameter, dtc_write_xenpv_conf will add as
many cdrom devices as it finds .iso files in the /var/lib/dtc\-xen/ttyssh_home/xenXX
folder.

.SH EXAMPLE

The following command:

dtc_write_xenpv_conf 14 512 '1.2.3.4 1.2.3.5' mYvNcPaSs boot.iso

will produce this config file in /etc/xen/xen14:

kernel = "/usr/lib/xen/boot/hvmloader"
.br
builder = 'hvm'
.br
memory = 512
.br
name = "xen14${VPSNAME}"
.br
vcpus=1
.br
pae=0
.br
acpi=0
.br
apic=0
.br
vif = [ 'type=ioemu, mac=00:00:00:00:00:00, ip=1.2.3.4 1.2.3.5' ]
.br
disk=[ 'phy:/dev/mapper/lvm1-xen14,ioemu:hda,w', 'file:/var/lib/dtc-xen/ttyssh_home/xen14/boot.iso,hdb:cdrom,r', 'file:/var/lib/dtc-xen/ttyssh_home/xen14/cd2.iso,hdc:cdrom,r' ]
.br
cdrom="/var/lib/dtc-xen/ttyssh_home/xen14/boot.iso"
.br
boot="d"
.br
vfb = [ "type=vnc,vncdisplay=21,vncpasswd=mYvNcPaSs" ]
.br
nographic=0
.br
vnc=1
.br
stdvga=1
.br
serial='pty'

Note that the above mac address will be calculated and depend on your node
name: it will NOT be like the above example.

.SH UPDATES

See
.B http://www.gplhost.com/software\-dtc\-xen.html
for updates.


doc/changelog
dtc-xen (0.2.7) unstable; urgency=low

  * Write RPM version including initscripts, specfile, autoconfiguration, SSL cert gen
  * Included daemon.py from http://www.clapper.org/software/python/daemon/ (BSD license)
  * Added the ability to self-daemonize
  * Added the respective documentation in the man pages

 -- Manuel Amador (Rudd-O) <rudd-o@rudd-o.com>  Fri, 12 Jun 2006 18:40 -0500

dtc-xen (0.2.6) unstable; urgency=low

  * Make it debian standard

 -- Thomas Goirand <thomas@goirand.fr>  Fri, 10 Nov 2006 21:12:00 +0800

dtc-xen (0.2.5) unstable; urgency=low

  * Add getNetworkUsage for each VPS

 -- Damien Mascord <tusker@tusker.org>  Wed, 25 Oct 2006 13:50:00 +0800

dtc-xen (0.2.4) unstable; urgency=low

  * Bugfixed the changeBSDKernel function to setup the xen startup

 -- Thomas GOIRAND <thomas@goirand.fr>  Sun, 17 Sep 2006 13:21:00 +0800

dtc-xen (0.2.3) unstable; urgency=low

  * Added anti-spoofing facility (IP addresses in the vif=)

 -- Thomas GOIRAND <thomas@goirand.fr>  Mon, 14 Sep 2006 13:21:00 +0800

dtc-xen (0.2.2) unstable; urgency=low

  * Added a physical console check to be sure nobody sends a param

 -- Thomas GOIRAND <thomas@goirand.fr>  Mon, 30 Aug 2006 11:21:00 +0800

dtc-xen (0.2.1) unstable; urgency=low

  * Added Xen version checking so getVPSState() always work

 -- Thomas GOIRAND <thomas@goirand.fr>  Mon, 28 Aug 2006 13:21:00 +0800

dtc-xen (0.2.0) unstable; urgency=low

  * Now all the functions we needed are working for the fist time

 -- Thomas GOIRAND <thomas@goirand.fr>  Tue, 15 Aug 2006 15:21:00 +0200

dtc-xen (0.1.2) unstable; urgency=low

  * Added packaging of the soap server

 -- Thomas GOIRAND <thomas@goirand.fr>  Sat, 22 Jul 2006 15:21:00 +0200

dtc-xen (0.1.1) unstable; urgency=low

  * Added the dtc_create_vps.sh script in /usr/sbin

 -- Thomas GOIRAND <thomas@goirand.fr>  Sat, 8 Jul 2006 16:06:38 +0200

dtc-xen (0.1.0) unstable; urgency=low

  * Debian users can now do dpkg-buildpackage -rfakeroot -us -uc

 -- Thomas GOIRAND <thomas@goirand.fr>  Tue, 28 Jun 2005 16:06:38 +0200


src/dtc_write_xenpv_conf.sh
#!/bin/sh

set -e

USAGE="Usage: $0 <xen id> <ram size> <allipaddrs> <vncpassword> <howtoboot>
Where allipaddrs is of the form '1.2.3.4 1.2.3.5' (eg: separated by space),
and howtoboot is 'name.iso' in /var/lib/dtc-xen/ttyssh_home/xenXX or 'hdd'"
if [ $# -lt 5 ]; then 
	echo $USAGE
fi

# Source the configuration in the config file!
. /etc/dtc-xen/dtc_create_vps.conf.sh

if [ "$LVMNAME" = "" ] ; then LVMNAME=lvm1 ; fi
VPSGLOBPATH=${VPS_MOUNTPOINT}
INSTALL_KERNELPATH="/boot/netbsd-INSTALL_XENU"
NORMAL_KERNELPATH="/boot/netbsd-XENU"

# Get parameters from command line
VPSNUM=$1
VPSNAME=xen${VPSNUM}
VPSHOSTNAME=xen${NODE_NUM}${VPSNUM}
RAMSIZE=$2
ALL_IPADDRS=$3
VNC_PASSWORD=$4
HOW_TO_BOOT=$5

calcMacAddr () {
	CHARCNT=`echo -n ${NODE_NUM} | wc -m`
	if [ ""${CHARCNT} = "5" ] ; then
		MINOR_NUM=`echo ${NODE_NUM} | awk '{print substr($0,4,2)}'`
		MAJOR_NUM=`echo ${NODE_NUM} | awk '{print substr($0,2,2)}'`
		MEGA_NUM=`echo ${NODE_NUM} | awk '{print substr($0,1,1)}'`
	else
		MINOR_NUM=`echo ${NODE_NUM} | awk '{print substr($0,3,2)}'`
		MAJOR_NUM=`echo ${NODE_NUM} | awk '{print substr($0,1,2)}'`
		MEGA_NUM="0"
	fi
	MAC_ADDR=`echo 00:00:2$MEGA_NUM:$MAJOR_NUM:$MINOR_NUM:$VPSNUM`
}
calcMacAddr

echo -n "kernel = \"/usr/lib/xen/boot/hvmloader\"
builder = 'hvm'
memory = ${RAMSIZE}
name = \"${VPSNAME}\"
vif = [ 'type=ioemu, mac=${MAC_ADDR}, ip=${ALL_IPADDRS}' ]
disk=[ 'phy:/dev/mapper/${LVMNAME}-xen${VPSNUM},ioemu:hda,w'" >/etc/xen/${VPSNAME}
# Set the additional cdrom drives: add all *.iso files to the config file
HDDLIST="bcdefghijklmnopqrstuvwxyz"
INCREMENT=2
for i in `find /var/lib/dtc-xen/ttyssh_home/xen${VPSNUM}/ -mindepth 1 -maxdepth 1 -iname '*.iso' | cut -d'/' -f7 | tr \\\r\\\n ,\ ` ; do
	DRIVE_LETTER=`echo ${HDDLIST} | awk '{print substr($0,'${INCREMENT}',1)}'`
	INCREMENT=$(( $INCREMENT + 1))
	echo -n ,\'file:/var/lib/dtc-xen/ttyssh_home/xen${VPSNUM}/$i,hd${DRIVE_LETTER}:cdrom,r\' >>/etc/xen/${VPSNAME}
done
echo " ]" >>/etc/xen/${VPSNAME}

# Set the VNC configuration
if [ -z "${VNC_PASSWORD}" -o "${VNC_PASSWORD}" = "none" ] ; then
	echo "nographic=1
vnc=0" >>/etc/xen/${VPSNAME}
else
	echo "vfb = [ \"type=vnc,vncdisplay=21,vncpasswd=${VNC_PASSWORD}\" ]
nographic=0
vnc=1
stdvga=1" >>/etc/xen/${VPSNAME}
fi

# Set the boot device
if [ ! "${HOW_TO_BOOT}" = "hdd" -a -e /var/lib/dtc-xen/ttyssh_home/xen${VPSNUM}/${HOW_TO_BOOT} ] ; then
	echo "cdrom=\"/var/lib/dtc-xen/ttyssh_home/xen${VPSNUM}/${HOW_TO_BOOT}\"
boot=\"d\"" >>/etc/xen/${VPSNAME}
else
	echo "boot=\"c\"
nographic=1" >>/etc/xen/${VPSNAME}
fi
echo "serial='pty'" >>/etc/xen/${VPSNAME}
